web3/providers/async_base.py

```diff
@@ -66,6 +71,28 @@
     )
 
 
+class SupportsBatching(Protocol):
```

Comment:
This works well. It makes me think maybe we should have a base class for that all providers originally stem from before deviating into sync + async behavior, that houses all the shared attributes - just like we do with `BaseWeb3`, `BaseEth`, etc... Then we wouldn't need to create this ghost type.

That's a stray thought for another time. This looks good 👍🏼 

My one comment would be to move this and the sync decorator into `web3._utils.batching`, that way we only define these types and the `SupportsBatching` class once and share it. Plus, that would be a better place to house this logic imo.